waiter/src/waiter/state.clj

```diff
@@ -46,6 +53,37 @@
   (and (zero? slots-used)
        (not-any? #(contains? status-tags %) [:killed :locked])))
 
+(defn- find-max
+  "Uses the provided comparison function to find the max element in collection."
+  [comparison-fn collection]
+  (reduce #(if (pos? (comparison-fn %1 %2)) %1 %2) (first collection) (rest collection)))
+
+(defn- only-lingering-requests?
+  "Retrurns truthy value if the instance is idle or processing expired requests based on the earliest-request-threshold-time"
+  [request-id->use-reason-map earliest-request-threshold-time]
+  (every? #(t/before? % earliest-request-threshold-time)
+          (->> request-id->use-reason-map
+               vals
+               (map :time))))
+
+(defn find-oldest-acceptable-expired-instance
+  "Retrieves the _oldest_ acceptable expired instance. An expired instance is acceptable if:
+   1. it passes the acceptable-instance-id? predicate, and
+   2. it is idle or all its requests are older than lingering-request-threshold-ms ms.
+   We do not explicitly check whether instances are idle as `find-killable-instance` should handle that scenario."
```

Comment:
This function would never return an idle instance, so this comment may be confusing.